src/relay/blockchain/events_informations.py

```diff
@@ -0,0 +1,133 @@
+import math
+from collections import namedtuple
+from typing import List
+
+from relay.blockchain.currency_network_events import (
+    BalanceUpdateEvent,
+    BalanceUpdateEventType,
+    TrustlineUpdateEventType,
+)
+from relay.network_graph.interests import calculate_interests
+
+InterestAccrued = namedtuple("InterestAccrued", ["value", "interest_rate", "timestamp"])
+
+DIRECTION_SENT = "sent"
+DIRECTION_RECEIVED = "received"
+
+
+def get_list_of_paid_interests_for_trustline(
+    events_proxy, currency_network_address, user, counterparty
+) -> List[InterestAccrued]:
+    """Get all balance changes of a trustline because of interests and the time at which it occurred.
+    Assumes the trustline exists"""
+
+    balance_update_events = events_proxy.get_trustline_events(
+        currency_network_address, user, counterparty, event_name=BalanceUpdateEventType
+    )
+
+    timestamps = [
+        balance_update_event.timestamp for balance_update_event in balance_update_events
+    ]
+
+    balances = get_sorted_balances_from_update_events(balance_update_events)
+    interest_rates = [
+        get_interests_rates_of_trustline_for_user_before_timestamp(
```

Comment:
The call to `get_interests_rates_of_trustline_for_user_before_timestamp` will ask for the same TrustlineUpdateEvents for each balance update event that you have. That slows down the computation and makes the code harder to understand.
Please write this in a way that you have a function (or maybe even better a class) that does the computation on pure data without making any network or database calls. From what I can tell you only need to provide the BalanceUpdate and TrustlinesUpdate events.

Reply:
I changed it as you suggested.

Reply:
> I changed it to get all necessary events at the beginning of the function get_list_of_paid_interests_for_trustline and pass the events to utility functions. 

good, but there's now a natural point to extract the computation that happens afterwards into it's own function function. So, please split the function. 

What happens before the computation could be made a method in the events_proxy class. But I'm also fine with keeping it as a function.

> We could further optimize get_interests_rates_of_trustline_for_user_before_timestamp

I think that's not necessary. Let's keep this simple.